routing/DELEGATED_CONTENT_ROUTING_HTTP.md

```diff
@@ -0,0 +1,253 @@
+# Delegated Content Routing HTTP API
+
+![wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) Delegated Content Routing HTTP API
+
+**Author(s)**:
+- Gus Eggert
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+"Delegated content routing" is a mechanism for IPFS implementations to use for offloading content routing to another process/server. This spec describes an HTTP API for delegated content routing.
+
+# API Specification
+The Delegated Content Routing Routing HTTP API uses the `application/json` content type by default. 
+
+As such, human-readable encodings of types are preferred. This spec may be updated in the future with a compact `application/cbor` encoding, in which case compact encodings of the various types would be used.
+
+## Common Data Types:
+
+- CIDs are always string-encoded using a [multibase](https://github.com/multiformats/multibase)-encoded [CIDv1](https://github.com/multiformats/cid#cidv1).
+- Multiaddrs are string-encoded according to the [human-readable multiaddr specification](https://github.com/multiformats/multiaddr#specification)
+- Peer IDs are string-encoded according [PeerID string representation specification](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#string-representation)
+- Multibase bytes are string-encoded according to [the Multibase spec](https://github.com/multiformats/multibase), and *should* use Base64.
+- Timestamps are Unix millisecond epoch timestamps
+
+Until required for business logic, servers should treat these types as opaque strings, and should preserve unknown JSON fields.
+
+### Versioning
+This API uses a standard version prefix in the path, such as `/v1/...`. If a backwards-incompatible change must be made, then the version number should be increased.
+
+### Provider Records
+A provider record contains information about a content provider, including the transfer protocol and any protocol-specific information useful for fetching the content from the provider.
+
+The information required to write a record to a router (*"write" provider records*) may be different than the information contained when reading provider records (*"read" provider records*).
+
+For example, indexers may require a signature in `bitswap` write records for authentication of the peer contained in the record, but the read records may not include this authentication information.
+
+Both read and write provider records have a minimal required schema as follows:
+
+```json
+{
+    "Protocol": "<transfer_protocol_name>",
+    ...
+}
+```
+
+where:
+
+- `Protocol` is the name of a transfer protocol
+  - these typically map to multicodec codes (such as bitswap => 2304)
+  - these don't use multicodec codes directly to allow for versioning of provider records, e.g. to allow for "bitswap-v2" provider records (which would also map to 2304 code but with different schema/semantics)
```

Comment:
We change this to use a "custom" string instead of using multicodecs for two reasons:
- Having a number as the identification for protocols was not super user-friendly. Example: using the string `"2304"` instead of `"bitswap"`. We need to use a string on that parameter to be able to follow OpenAI specs for [discriminator objects](https://spec.openapis.org/oas/latest.html#discriminator-object).
- Multiformat names are not intended to be unique, so we didn't want to break our API using multiformat names. Example: https://github.com/multiformats/multicodec/pull/283

Reply:
Basically the goal here is to provide some protection against "what if we get the schema for <transfer protocol> wrong?". It doesn't make sense to add "bitswap-v2" to the multicodec table just because we got the schema wrong in the content routing HTTP API--nothing about the bitswap protocol itself would have changed.

So this defines its own table of schema names that are scoped to the content routing HTTP API, and clients/servers can opt-in to any subset of that table if they wish.

> These records are quickly moving towards a double-hashed variation where content routers like the indexers are not going to be able to re-format the records themselves

This seems like the most important issue here. If the routers can't interpret records then how will they support `transfer` and `transport` query params? I'm not up-to-speed on the double-hashing proposal so my knowledge is limited here, but is there a way to change it so that routers have more information in the record so they can do smarter routing?

Reply:
Talked offline w/ @masih about this, he had a good suggestion to define this as a multicodec id and have a separate field for the schema version, I'll update this spec w/ that as it solves the problem elegantly.

Concretely my concern is around coupling the HTTP API spec w/ the multicodec transfer protocol. Concrete example: we are currently talking about adding transfer protocols to the DHT too, so that kubo can properly support different transfer protocols. If that is also based on the same multicodec id, then an HTTP API change would bump the version and those records would not be consumable by any nodes in the network, even nodes that use the DHT instead of HTTP. Also it would require record publishers to be upgraded to produce the new multicodec (Filecoin SPs). Requiring upgrades of the entire network just b/c we made a change to the HTTP schema here seems like an unreasonably high cost. I want to decouple the two so that we can make HTTP schema changes but still end up w/ the same multicodec id so the entire network doesn't need to upgrade. I was thinking of keeping a separate table for this, but I think Masih's approach works too and is more elegant since you can reuse the multicodec table for interpreting the transfer protocol.

Reply:
> This seems like the most important issue here. If the routers can't interpret records then how will they support transfer and transport query params?

I'm curious about the plan here too. Particularly as it relates to say supporting multiple different bitswap versions (e.g. querying for Bitswap 1.2 support), or perhaps a code called "named" which has name + data fields and would let us reference new protocols without registering in the global table (e.g. as is the case with libp2p protocols today).

Reply:
Fwiw we already have [protocol names and versions in libp2p identify protocols list](https://github.com/libp2p/specs/tree/master/identify#protocols) – perhaps just refer that spec and say "use the same strings"?

```console
$ ipfs id | jq .Protocols                                                                                                                   ~
[
  "/floodsub/1.0.0",
  "/ipfs/bitswap",
  "/ipfs/bitswap/1.0.0",
  "/ipfs/bitswap/1.1.0",
  "/ipfs/bitswap/1.2.0",
  "/ipfs/id/1.0.0",
  "/ipfs/id/push/1.0.0",
  "/ipfs/kad/1.0.0",
  "/ipfs/lan/kad/1.0.0",
  "/ipfs/ping/1.0.0",
  "/libp2p/autonat/1.0.0",
  "/libp2p/circuit/relay/0.1.0",
  "/libp2p/circuit/relay/0.2.0/hop",
  "/libp2p/circuit/relay/0.2.0/stop",
  "/libp2p/dcutr",
  "/libp2p/fetch/0.0.1",
  "/meshsub/1.0.0",
  "/meshsub/1.1.0",
  "/p2p/id/delta/1.0.0",
  "/x/"
]
```